yastd/lib/yast2/installation_progress.rb

```diff
@@ -22,23 +22,70 @@
 # YaST specific code lives under this namespace
 module Yast2
   # This class represents the installer status
-  class Progress
-    def initialize(dbus_obj)
+  class InstallationProgress
+    KNOWN_STEPS = 3 # keep it in sync with installer.rb
+    def initialize(dbus_obj, logger: nil)
       @dbus_obj = dbus_obj
+      @logger = logger
       @total_pkgs = 0
       @remaining_pkgs = 0
     end
 
+    def partitioning(&block)
+      report(
+        # TODO: localization
+        "Partitioning target disk", 0
+      )
+      block.call(self)
+      report(
+        # TODO: localization
+        "Partitioning finished", 1
+      )
+    end
+
+    def package_installation(&block)
+      report(
+        # TODO: localization
+        "Starting to install packages", 1
+      )
+      block.call(self)
+      report(
+        # TODO: localization
+        "Package installation finished", 2
+      )
+    end
+
+    def bootloader_installation(&block)
+      report(
+        # TODO: localization
+        "Starting to deploy bootloader", 2
+      )
+      block.call(self)
+      report(
+        # TODO: localization
+        "Installation finished", 3
+      )
+    end
+
     def packages_to_install=(value)
       @total_pkgs = @remaining_pkgs = value
     end
 
     def package_installed
       @remaining_pkgs -= 1
-      @dbus_obj&.report_progress(
+      report(
         # TODO: localization
         "Installing packages (#{@remaining_pkgs} remains)",
-        1, 0, @total_pkgs, @total_pkgs - @remaining_pkgs
+        1, substeps: @total_pkgs, current_substep: @total_pkgs - @remaining_pkgs
+      )
+    end
+
+  private
+
+    def report(title, step, substeps: 0, current_substep: 0)
```

Comment:
Why not making it public and use it directly in the installer.rb?

I mean,

```ruby

...

def install
  ...
  Yast::Installation.destdir = "/mnt"
  # Imaging we have a collection of steps or so where partitioning could be...
  # 
  # class Partitioning < InstallationStep
  #   def start_label
  #     _("Partitioning target disk")
  #   end
  # 
  #   def perform
  #     Yast::WFM.CallFunction("inst_prepdisk", [])
  #   end
  #  
  #   ...
  steps = [partitioning, package_installation, bootloader_installation]

  # So...

  progress = InstallationProgress.new(@dbus_obj, steps.size, logger: logger)
  
  steps.each_with_index |step, index| do
    progress.report(step.start_label, index)
    step.perform
    progress.report(step.end_label, index + 1)
  end

  ...
end

...

```

Reply:
well, I am still not sure whose responsibility it is. Also progress needs to know about number of steps so it can report it. So question is who owns responsibility for proper messaging. Still not sure what would be the best solution

Reply:
> Also progress needs to know about number of steps so it can report it

Right. That's why I put the "steps.size" in the `InstallationProgress` constructor. To tell it the number of steps when creating it, instead of having an internal constant that must be in sync with the installer.rb.

Anyway, let's continue playing around. Thank you!